### PR TITLE
Adjust `polymorphicInstance()`

### PR DESCRIPTION
**Pull request by cameel**
The PR addresses https://github.com/ethereum/solidity/pull/14510#discussion_r1363755155 and https://github.com/ethereum/solidity/pull/14510#discussion_r1363752412.

We discussed this last week on the call - `unifyGeneralized()` was a part of some idea that was not fully realized in the end and all we really need is just a call to `TypeEnvironment::fresh()`. That's what `polymorphicInstance()` does now.

The PR also adds some tests covering the uses of `fresh()` and `polymorphicInstances()`. This shows that the PR does not break things both here and with my subsequent changes in #14606.

### CODE REVIEW

File: test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+function f(x, y: X, z: U(Y)) {}
+
+function run(a: T, b: U(T), c: U(U(T))) {
+    let fRef = f;
+
+    // FIXME: The second call does not unify. Apparently the call via a function pointer does not
+    // get fresh type variables like a direct call.
+    fRef(a, a, b);
+    fRef(b, b, c);
```
Comment (cameel) [inline test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol]: This currently does not work. Type variables in `fRef` won't unify with both calls. Perhaps this was missing a `polymorphicInstance()` in the implementation?

I can address it with a separate PR after I'm done with the current work on fixed type variables. Or maybe I should just start creating issues for these things?

Comment (ekpyron) [reply test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol]: This *should not* unify. This is exactly what it means to *not* have let-polymorphism.

Comment (ekpyron) [reply test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol]: There is no way to fix this other than *introducing* let-polymorphism in the first place, which would complicate the entire mechanism significantly :-).

**Comment by ekpyron** [comment]
The funny thing is: the last thing I did before pushing the prototype was to do the exact opposite :-) - i.e. to introduce ``polymorphicInstance`` in the first place in order to isolate the occurrances of ``fresh()`` to make it easier to replace them by a mechanism accounting for https://github.com/ethereum/solidity/pull/14606

File: test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol (hunk 1 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol]: ok then. I'll remove the FIXME :)

**Comment by cameel** [comment]
Why would it be any easier to replace `polymorphicInstance()` than to replace `fresh()`?

**Comment by ekpyron** [comment]
> Why would it be any easier to replace `polymorphicInstance()` than to replace `fresh()`?

The point was to collect all instances of refreshening for the purpose of generating a polymorphic schema of a global definition and then checking what's left.

Also: producing a freshened type is a technical operation on the type - producing a polymorphic instance of a global type scheme is a semantic description of what's happening.

**Comment by ekpyron** [comment]
> Also: producing a freshened type is a technical operation on the type - producing a polymorphic instance of a global type scheme is a semantic description of what's happening.

Based on that I'd actually not remove it. (Removing the entire function that gives the operation that happens here a descriptive name is quite different from removing the additional type variable it unnecessarily introduces :-))

File: test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol (hunk 1 above)
Comment (cameel) [reply test/libsolidity/syntaxTests/experimental/inference/polymorphic_function_call_via_variable.sol]: Done.

**Comment by cameel** [comment]
`polymorphicInstance()` restored.

**Review by ekpyron — APPROVED** [review]

**Comment by cameel** [comment]
I need reapproval. Only pushed a small stylistic tweak.

**Comment by ekpyron** [comment]
Before we merge new changes, we first need to clean up the mess in the base branch.

**Comment by cameel** [comment]
https://github.com/ethereum/solidity/issues/14570#issuecomment-1788884982

> It may actually make sense to just rename ``polymorphicInstance`` to ``generalize``.

I wanted to merge this already, but in that case maybe I should still change it before I do?

**Comment by ekpyron** [comment]
> I wanted to merge this already, but in that case maybe I should still change it before I do?

No hurry, just bringing up the option.

File: libsolidity/experimental/analysis/TypeInference.cpp [hunk 2]
```diff
@@ -481,7 +480,10 @@ experimental::Type TypeInference::handleIdentifierByReferencedDeclaration(langut
 		else if (dynamic_cast<FunctionDefinition const*>(&_declaration))
 			return polymorphicInstance(*declarationAnnotation.type);
 		else if (dynamic_cast<TypeClassDefinition const*>(&_declaration))
-			return polymorphicInstance(*declarationAnnotation.type);
+		{
+			solAssert(TypeEnvironmentHelpers{*m_env}.typeVars(*declarationAnnotation.type).empty());
```
Comment (ekpyron) [inline libsolidity/experimental/analysis/TypeInference.cpp:484]: Wait, what? That doesn't seem right - shouldn't there be exactly one variable in there?

Comment (ekpyron) [reply libsolidity/experimental/analysis/TypeInference.cpp:484]: Ah, no, this is term context, so the use of the type class name as a special concrete type with its functions attached as members... I hope the distinction will become clearer after the upcoming refactors, we'll see.

Comment (cameel) [reply libsolidity/experimental/analysis/TypeInference.cpp:484]: So fine with merging the PR?

**Review by ekpyron — APPROVED** [review]

**Pull request merged**